csrc/id_graphs.cpp

```diff
@@ -0,0 +1,3620 @@
+  for (auto iel_expr : iel_stmt_sort.exprs()) {
+    IdGroups input_groups = intersection_exact_loop_graph.inputGroups(iel_expr);
+
+    // Check if any inputs need promotion indicating this expr group needs to
+    // be replayed with promoted inputs
+    std::vector<IterDomain*> promoted_inputs;
+    bool an_input_was_promoted = false;
+
+    for (auto inp : input_groups) {
+      auto inp_promo_it = iel_promotion_map.find(inp);
+      if (inp_promo_it == iel_promotion_map.end()) {
+        promoted_inputs.push_back(inp->front());
+      } else {
+        promoted_inputs.push_back(inp_promo_it->second);
+        an_input_was_promoted = true;
+      }
+    }
+
+    if (!an_input_was_promoted) {
+      // No inputs need promotion so just continue
+      continue;
+    }
+
+    Expr* replay = nullptr;
+
+    auto promoted_input_groups = intersection_exact_loop_graph.toGroups(
+        VectorOfUniqueEntries<IterDomain*>{
+            promoted_inputs.begin(), promoted_inputs.end()});
+
+    // Before replaying, check if there's already an expression like this, if so
+    // use that for promotion. We would need the iel entries for non-promoted
+    // inputs to match exactly to reuse the expression.
+    ExprGroups non_promoted_input_uses;
+    for (auto iel_group : promoted_input_groups.intersect(input_groups)) {
+      non_promoted_input_uses.pushBack(
+          intersection_exact_loop_graph.uniqueUses(iel_group));
+    }
+
+    for (auto iel_use_group : non_promoted_input_uses) {
+      if (transformAtributesMatch(iel_expr->front(), iel_use_group->front())) {
+        auto use_inps =
+            ir_utils::filterByType<IterDomain>(iel_use_group->front()->inputs())
+                .vector();
+        bool inps_match = true;
+        for (auto inp_i : c10::irange(use_inps.size())) {
+          inps_match = inps_match &&
+              intersection_exact_loop_graph.disjointIdSets().strictAreMapped(
+                  use_inps[inp_i], promoted_inputs[inp_i]);
+        }
+        if (inps_match) {
+          replay = iel_use_group->front();
+          break;
+        }
+      }
+    }
+
+    if (replay == nullptr) {
+      replay = addReplayAs(promoted_inputs, iel_expr->front());
+      std::cout << "  ***REPLAY***:\n    " << iel_expr->front()
+                << "    As:" << replay->toString();
+    } else {
+      std::cout << "  Matched replay found: " << replay->toString();
+    }
+
+    auto out_groups = intersection_exact_loop_graph.outputGroups(iel_expr);
+
+    // Mark outputs as having a promoted iter domain
+    auto replay_out_ids =
+        ir_utils::filterByType<IterDomain>(replay->outputs()).vector();
+
+    TORCH_INTERNAL_ASSERT(replay_out_ids.size() == out_groups.size());
+
+    for (auto i : c10::irange(replay_out_ids.size())) {
+      iel_promotion_map[out_groups.vector()[i]] = replay_out_ids[i];
+    }
+  }
+
+  // Map from an exact iter domain group, to all the exact iter domain groups it
```

Comment:
Would be helpful to have a clear definition of what covering means.

Reply:
Inputs of, but also considering rfactor iter domains as inputs